docs: update conventions to latest Move features (#20791)
MystenLabs/sui: docs/content/concepts/sui-move-concepts/conventions.mdx

// === View Functions ===

// === Admin Functions ===

// === Package Functions ===

// === Private Functions ===

// === Test Functions ===
```

Here, "Public Functions" are the functions modifying state, "View Functions" are often on-chain getters or off-chain helpers. The latter are not necessary because data is easily readable by querying objects. `init` function should be the first function in the module if it exists.

Try to sort your functions by their purpose and according to the user flow to improve readability, you can also use explicit function names like `admin_set_fees` to make it clear what the function does.

Ideally, test functions should only consists of `[test_only]` helpers for the actual tests that should be located in the tests directory.

Imports can be grouped by dependency, for example:

```move
use std::string::String;
use sui::{
    coin::Coin,
    balance,
    table::Table
};
use my_dep::battle::{Battle, Score};
```

## Naming Conventions

### Constants

Constants should be in uppercase and use snake case. Errors are specific constants that use PascalCase and start with an E. Make them descriptive.

```move
module conventions::constants;

> const MAX_NAME_LENGTH: u64 = 64;

Reviewer: Might be useful to add a comment here clarifying that this is a non-error constant (so upper snake case good), and that the two examples below are error consts

// this is right
const EInvalidName: u64 = 0;
// this is wrong
const E_INVALID_NAME: u64 = 0;
```

### Structs

To anticipate for future features, struct must be declared with the `public` modifier. Abilities must always be declared in this order: `key`, `copy`, `drop`, `store`. Do not use 'potato' in the name of structs. The lack of abilities define it as a potato pattern.

Structs support positional fields that can be used for simple wrappers, dynamic fields keys or as tuples.

> Some the structs will be used to emit events, these structs should be named with the `Event` suffix.

Reviewer 2: ```suggestion
Use the `Event` suffix to name structs that emit events.
```
Conciseness

```move
module conventions::request;

// dynamic field keys
public struct ReceiptKey(ID) has copy, drop, store;

// dynamic field
public struct Receipt<Data> has key, store {
    id: UID,
    data: Data
}

// right naming

> public struct Request {}

Reviewer: I think we'd have a weak preference for `public struct Request()` for this now that this is valid syntax.

// wrong naming
public struct RequestPotato {}
```

> ### CRUD functions names

Reviewer 2: ```suggestion
### CRUD function names
```

> The functions below follow standard CRUD (Create, Read, Update, Delete) naming conventions:

Reviewer 2: ```suggestion
The following functions follow standard CRUD (Create, Read, Update, Delete) naming conventions:
```
Avoid 'below' and 'above' words because you never can be sure of how the final render will appear.

- `new`: Creates an empty object.
- `empty`: Creates an empty struct.
- `create`: Creates an initialized object or struct.
- `add`: Adds a value.
- `remove`: Removes a value.
- `exists`: Checks if a key exists.
- `contains`: Checks if a collection contains a value.
- `borrow`: Returns an immutable reference of a struct or object.
- `borrow_mut`: Returns a mutable reference of a struct or object.
- `property_name`: Returns an immutable reference or a copy of a field.
- `property_name_mut`: Returns a mutable reference of a field.
- `drop`: Drops a struct.
- `destroy`: Destroys an object or data structure that has values with the **drop** ability.
- `destroy_empty`: Destroys an empty object or data structure that has values with the **drop** ability.

> - `to_name`: Transforms an Type X to Type Y.

Reviewer 2: ```suggestion
- `to_name`: Transforms a Type X to Type Y.
```

> - `from_name`: Transforms an Type Y to Type X.

Reviewer 2: ```suggestion
- `from_name`: Transforms a Type Y to Type X.
```

### Generics 

> Generics can be declared using single letter names or full names. By convention, developers use T and U for generic types, but you can use a more descriptive name if it is not confusing with other types. Always prioritize readability. 

Reviewer 2: ```suggestion
Declare generics using single letter names or full names. By convention, developers use `T` and `U` for generic types, but you can use a more descriptive name if it is not confusing with other types. Always prioritize readability. 
```

```move
module conventions::generics;

// single letter name
public struct Receipt<T> has store { ... }

// full name
public struct Receipt<Data> has store { ... }
```

## Code Structure

> 

Reviewer 2: Stacked titles.

public fun new(deposit: Coin<SUI>, ctx: &mut TxContext): OwnedWallet {
    // Implementation omitted.
    abort(0)
}

// If you wish to add extra functionality to a vesting wallet, it is best to share the object.
// For example, if you wish the issuer of the wallet to be able to cancel the contract in the future.
public fun new_shared(deposit: Coin<SUI>, ctx: &mut TxContext) {
    // Implementation omitted.
    // shares the `SharedWallet`.
    abort(0)
}
```

### Admin capability

In admin-gated functions, the first parameter should be the capability. It helps the autocomplete with user types.

```move
module conventions::social_network;

use std::string::String;

public struct Account has key {
    id: UID,
    name: String
}

public struct Admin has key {
    id: UID,
}

// right -> cap.update(&mut account, b"jose".to_string());
public fun update(_: &Admin, account: &mut Account, new_name: String) {
    // Implementation omitted.
    abort(0)
}

// wrong -> account.update(&cap, b"jose".to_string());
public fun set(account: &mut Account, _: &Admin, new_name: String) {
    // Implementation omitted.
    abort(0)
}
```

## Documentation

> There is nothing more pleasant than a well-written and well-documented codebase. If some support the fact that clean code is self-documenting, I would say that well-documented code is self-explanatory.

Reviewer 2: ```suggestion
There is nothing more pleasant than a well-written and well-documented codebase. If some support the idea that clean code is self-documenting, well-documented code is self-explanatory.
```
There's no 'I' in 'docs' ;)

### Comments

Document your code by explaining functions and structs in simple terms using the `///` syntax (doc comment). If you want to add technical insights for developers that might use your code, use the `//` syntax (regular comment).

> Field comments should be used to describe the properties of your structs. In complex functions, you can also describe the parameters and return values.

Reviewer 2: ```suggestion
Use field comments to describe the properties of your structs. In complex functions, you can also describe the parameters and return values.
```

```move
module conventions::hero;

use std::string::String;
use sui::kiosk::{Kiosk, KioskOwnerCap};

public struct Hero has key, store {
    id: UID,
    // power of the nft
    power: u64
}

> /// creates and returns a new Hero object

Reviewer 2: ```suggestion
/// Creates and returns a new Hero object
```
For doc comments, should probably use sentence case.

public fun new(ctx: &mut TxContext): Hero {
    Hero {
        id: object::new(ctx),
        power: 0
    }
}

// should be initialized before being shared
public fun initialize_hero(hero: &mut Hero) {
    hero.power = 100;
}

public fun start_battle(
    self: &mut Kiosk, // user kiosk
    cap: &KioskOwnerCap, // user kiosk owner cap
    _policy: &TransferPolicy<Hero>, // transfer policy for the game
    hero_id: ID, // hero to use
    battle_id: String // id of the battle to start
) { 
    // Implementation omitted.
    abort(0)
}
```

### README

> The README file should be located in the root of the package and should be named `README.md`. It should contain a description of the package, the purpose of the package, and how to use it.

Reviewer 2: ```suggestion
Create a `README.md` file in the root of the package. Include a description of the package, the purpose of the package, and how to use it.
```
Conciseness.
